### PR TITLE
Improve memory usage and speed of SPARSE_HASHED/HASHED dictionaries

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve memory usage and speed of `SPARSE_HASHED`/`HASHED` dictionaries (e.g. `SPARSE_HASHED` now eats 2.6x less memory, and is ~2x faster)

### Documentation entry for user-facing changes
`SPARSE_HASHED` now uses a "packed" hash table for small structures (<= 16 bytes), which consumes less memory and works faster.

And you can now also configure `MAX_LOAD_FACTOR` (`[0.5, 0.9]`) for the `HASHED`/`SPARSE_HASHED` dictionary layouts, which potentially decreases their memory usage. My testing shows that even 0.99 load factor works well enough (faster then previous implementation of `SPARSE_HASHED`), while reducing memory usage up to 2x.

### TL;DR;

**Firstly**, this PR **switches `SPARSE_HASHED` layout to `HashMap`** (internal ClickHouse hash table), but **with packed structure** (to avoid wasting memory for padding). This is done only for small enough key-value pairs (otherwise `google::sparse_hash_map` outperforms `HashMap`)

Consider the following example <UInt64, UInt16> as <Key, Value>, but this pair will also have a 6 byte padding (on amd64), so this is almost **40% of space wastage**.

Another problem with sparsehash is that it is not very friendly to memory allocator (especially jemalloc, it does not like tons of small reallocs, due to slabs, redis even did some fragmentation to overcome this - https://github.com/redis/redis/pull/3720).

**Secondly**, it allows to **configure the load factor of the HASHED/SPARSE_HASHED** dictionaries now, as it turns out, HashMap/**PackedHashMap works great even with max load factor of 0.99**. By "great" I mean it least it works faster then google sparsehash, and not to mention it's friendliness to the memory allocator (it has zero fragmentation since it works with a continuous memory region).

Here are some numbers (for 1e9 elements and UInt16 value):

_(but likely I will update them, since they does not looks accurate to me)_

settings|load (sec)|read (sec)|read (million rows/s)|bytes_allocated|RSS
-|-|-|-|-|-
HASHED upstream|-|-|-|-|35GiB
SPARSE_HASHED upstream|-|-|-|-|26GiB
-|-|-|-|-|-
sparse_hash_map glibc hashbench|-|-|-|-|17.5GiB
sparse_hash_map packed allocator|4.32|101.878|231.48|-|17.7GiB
PackedHashMap 0.5|15.514|42.35|23.61|20GiB|22GiB
hashed 0.95|34.903|115.615|8.65|16GiB|18.7GiB
**PackedHashMap 0.95**|**93.6**|**19.883**|**10.68**|**10GiB**|**12.8GiB**
PackedHashMap 0.99|26.113|83.6|11.96|10GiB|12.3GiB

**As you can see PackedHashMap looks way more better then current implementation of SPARSE_HASHED: 2.6x less memory, and ~2x faster**

Refs: #6894
Refs: #40003

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit a4f03993205aa51f598a4610ce7f3fc7c66654a1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49380/a4f03993205aa51f598a4610ce7f3fc7c66654a1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by UnamedRus** [comment]
May be, it make sense to implement that for HASHED_ARRAYS layout?

> Consider the following example <UInt64, UInt16> as <Key, Value>, but this pair will also have a 6 byte padding (on amd64), so this is almost 40% of space wastage.

Improvement will be smaller (if we consider UInt64, UInt32), probably closer to 25%, but it still quite a bit.

**Comment by azat** [comment]
> May be, it make sense to implement that for HASHED_ARRAYS layout?

@UnamedRus It does not need this since it has `HashMap<UInt64, size_t>` which does not have padding (on x86_64 at least).

>Improvement will be smaller (if we consider UInt64, UInt32), probably closer to 25%, but it still quite a bit

For `<UInt64, **UInt32**>` - yes, but example has `<UInt64, **UInt16**>`

**Comment by UnamedRus** [comment]
> For <UInt64, **UInt32**> - yes, but example has <UInt64, **UInt16**>

It was in continuation for HASHED_ARRAYS question:) 

>  it has HashMap<UInt64, size_t> which does not have padding

But if we for example can fit it in 4 byte int, (4 bil of elements is quite a bit and should be enough for offset in array (and we theoretically can also have sharding here, so)), could it give some improvement or not?

**Comment by azat** [comment]
>But if we for example can fit it in 4 byte int, (4 bil of elements is quite a bit and should be enough for offset in array (and we theoretically can also have sharding here, so)), could it give some improvement or not?

Yes, but to do this some analog of `HASHED_ARRAY` need to be introduced that will allow `shards*UINT32_MAX` elements only.

**Comment by azat** [comment]
>[Performance Comparison [4/4] — 2 errors](https://s3.amazonaws.com/clickhouse-test-reports/49380/3daa1d880d4874993afc327f044f7587848aab6c/performance_comparison_[4_4]/report.html)

```
DB::Exception: Dictionary layout parameter value must be an UInt64 or String, got 'Float64' instead. Stack trace:
```

upstream does not suppport float type for dictionary parameters

>[Performance Comparison Aarch64 [4/4] — 2 errors, 1 too long](https://s3.amazonaws.com/clickhouse-test-reports/49380/3daa1d880d4874993afc327f044f7587848aab6c/performance_comparison_aarch64_[4_4]/report.html#test-times.group_by_sundy_li)

```
The test 'group_by_sundy_li' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up
```

It fails in [upstream too](https://play.clickhouse.com/play?user=play#c2VsZWN0IHRlc3Rfc3RhdHVzLCB0ZXN0X25hbWUsIHJlcG9ydF91cmwgZnJvbSBjaGVja3Mgd2hlcmUgcmVwb3J0X3VybCBsaWtlICclcGVyZiUnIGFuZCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gaW50ZXJ2YWwgMiB3ZWVrIGFuZCB0ZXN0X25hbWUgbGlrZSAnJWdyb3VwX2J5X3N1bmR5X2xpJSc=)

**Comment by azat** [comment]
@alexey-milovidov (since you merged the initial implementation in #6894) or maybe @kitaisreal can you take a look please? (I will rebase after review, to avoid wasting CI cycles)

**Comment by rschu1ze** [comment]
I can check if you don't mind.

**Comment by azat** [comment]
>I can check if you don't mind.

That would be great, thank you!

In the mean time I'd rebased on top upstream to fix conflicts (plus there were some check failures due to flaky tests)

**Comment by azat** [comment]
>[Stateless tests (debug, s3 storage) [6/6] — Tests are not finished, fail: 1, passed: 74, skipped: 2]https://s3.amazonaws.com/clickhouse-test-reports/49380/37d22bc33049e580b65976bbc78860f64cf2c2af/stateless_tests__debug__s3_storage__[6_6].html)

```
clickhouse-server: /build/src/Common/OpenTelemetryTraceContext.cpp:144: void DB::OpenTelemetry::SpanHolder::finish(): Assertion `current_thread_trace_context.span_id == span_id' failed.
```

>[Integration tests (asan) [1/6] — fail: 2, passed: 368, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/49380/37d22bc33049e580b65976bbc78860f64cf2c2af/integration_tests__asan__[1_6].html)

- `test_distributed_load_balancing` flaky - need to take a look

**Comment by azat** [comment]
@rschu1ze likely you remember about this PR, but just a friendly reminder (it case it got lost in mbox)

**Comment by rschu1ze** [comment]
Didn't forget about this PR, just got distracted by some other stuff. Review planned for the weekend or Monday at latest.

File: src/Dictionaries/HashedDictionaryCollectionTraits.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+namespace DB
+{
+
+/// sparse_hash_map/sparse_hash_set
+template <typename C>
+concept IsSparseHashTable = std::is_same_v<C, google::sparse_hash_map<
+    typename C::key_type,
+    typename C::mapped_type,
+    /// HashFcn is not exported in sparse_hash_map is public type
+    DefaultHash<typename C::key_type>>>;
```
Comment (rschu1ze) [inline src/Dictionaries/HashedDictionaryCollectionTraits.h:16]: Let's add some newlines between the `concept` definitions.

File: src/Dictionaries/HashedDictionaryCollectionTraits.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#include <type_traits>
+#include <sparsehash/sparse_hash_map>
+#include <Common/HashTable/Hash.h>
+
+namespace DB
+{
+
+/// sparse_hash_map/sparse_hash_set
+template <typename C>
+concept IsSparseHashTable = std::is_same_v<C, google::sparse_hash_map<
```
Comment (rschu1ze) [inline src/Dictionaries/HashedDictionaryCollectionTraits.h:12]: Looks to me that the term "SparseHashTable" is too loaded already. Maybe `IsGoogleSparseHashTable`?

File: src/Dictionaries/HashedDictionaryCollectionTraits.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    typename C::key_type,
+    typename C::mapped_type,
+    /// HashFcn is not exported in sparse_hash_map is public type
+    DefaultHash<typename C::key_type>>>;
+template <typename V>
+concept IsStdMapCell = requires (V v)
```
Comment (rschu1ze) [inline src/Dictionaries/HashedDictionaryCollectionTraits.h:18]: Slightly confusing. So `v` has `first` and `last`. How does that relate to being a "StdMapCell"?

File: src/Dictionaries/HashedDictionaryCollectionTraits.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+{
+    v->first;
+    v->second;
+};
+
+/// HashMap/HashSet and their Cells
+template <typename C>
+concept IsBuiltinHashTable = !IsSparseHashTable<C>;
```
Comment (rschu1ze) [inline src/Dictionaries/HashedDictionaryCollectionTraits.h:26]: I guess the concept will break if someone adds another external hash table in future. Maybe `IsBuildinHashTable` should check directly against the building hash table type.

File: src/Dictionaries/HashedDictionary.cpp [hunk 5]
```diff
@@ -1192,9 +1155,14 @@ void registerDictionaryHashed(DictionaryFactory & factory)
         if (shard_load_queue_backlog <= 0)
             throw Exception(ErrorCodes::BAD_ARGUMENTS,"{}: SHARD_LOAD_QUEUE_BACKLOG parameter should be greater then zero", full_name);
 
+        float max_load_factor = static_cast<float>(config.getDouble(config_prefix + dictionary_layout_prefix + ".max_load_factor", 0.5));
+        if (max_load_factor < 0.5 || max_load_factor > 0.99)
```
Comment (rschu1ze) [inline src/Dictionaries/HashedDictionary.cpp:1159]: (Ah, I asked about a sanity check for max_load_factor in another comment. Here it is.)

File: src/Common/HashTable/PackedHashMap.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <Common/HashTable/HashMap.h>
+
+/// A pair that does not initialize the elements, if not needed.
+///
+/// NOTE: makePairNoInit() is omitted for PackedPairNoInit since it is not
+/// required for PackedHashMap (see mergeBlockWithPipe() for details)
+template <typename First, typename Second>
+struct PackedPairNoInit
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Common/HashTable/PackedHashMap.h:10]: I was building locally and got this error:
```
In file included from /home/ubuntu/repo/ch4/src/Dictionaries/HashedDictionaryCollectionType.h:4:                                                                                                                               
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<128, unsigned int>>' [-Werror,-Wpacked]                             
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<256, unsigned int>>' [-Werror,-Wpacked]                             
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<128, int>>' [-Werror,-Wpacked]                                      
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<256, int>>' [-Werror,-Wpacked]                                      
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::Decimal<long>>' [-Werror,-Wpacked]                                            
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::Decimal<wide::integer<128, int>>>' [-Werror,-Wpacked]                         
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::Decimal<wide::integer<256, int>>>' [-Werror,-Wpacked]                         
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::DateTime64>' [-Werror,-Wpacked]                                               
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, StrongTypedef<wide::integer<128, unsigned int>, DB::UUIDTag>>' [-Werror,-Wpacked] 
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
fatal error: too many errors emitted, stopping now [-ferror-limit=]                                                                                                                                                            
20 errors generated.                                                                                                                                                                                                           
[46/1294] Building CXX object src/Parsers/CMa...rsers.dir/ASTCreateNamedCollectionQuery.cpp.o                                                                                                                                  
ninja: build stopped: subcommand failed.                                                                                                                                                                                       
```

[the remark this answers is not in the document]
Comment (azat) [reply src/Common/HashTable/PackedHashMap.h:10]: >I was building locally and got this error:

Yeah, I converted to clang-16 too early :)
The problem is that the interpretation of POD types (in terms of layout) had been changed since clang-16 - https://github.com/llvm/llvm-project/commit/277123376ce08c98b07c154bf83e4092a5d4d3c6

Will disable using of PackedHashMap for some types.

File: src/Dictionaries/HashedDictionaryCollectionType.h [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+/// google::sparse_hash_map.
+///
+/// Based on HashTableGrowerWithPrecalculation
+class alignas(64) HashTableGrowerWithMaxLoadFactor
+{
+    static constexpr size_t initial_size_degree = 8;
+    UInt8 size_degree = initial_size_degree;
+    size_t precalculated_mask = (1ULL << initial_size_degree) - 1;
+    size_t precalculated_max_fill = 1ULL << (initial_size_degree - 1);
+    float max_load_factor = 0.5;
```
[the remark this answers is not in the document]
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionType.h:60]: This cannot be done because of std::swap in HashTable

File: src/Dictionaries/HashedDictionaryCollectionTraits.h (hunk 4 above)
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionTraits.h:26]: I doubt that it worth it, since we will need too much code (almost 30 lines), but let's try.

File: src/Dictionaries/HashedDictionaryCollectionTraits.h (hunk 3 above)
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionTraits.h:18]: By std here it means any std compatible cell, like sparsehash uses. This is an `sparse_hashtable_iterator<std::pair>`, making it explicit will require more code, not sure that it worth it, since `->first` and `->second` likely means some `pair` anyway.

File: src/Dictionaries/HashedDictionaryCollectionType.h [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+    void setBufSize(size_t buf_size_)
+    {
+        size_degree = static_cast<size_t>(log2(buf_size_ - 1) + 1);
+        increaseSizeDegree(0);
+    }
+};
+static_assert(sizeof(HashTableGrowerWithMaxLoadFactor) == 64);
+
+///
```
[the remark this answers is not in the document]
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionType.h:127]: I will try to add some summary for this, but I thought that comments that exists in the file was enough (I have pretty verbose comments for useSparseHashForHashedDictionary and below)

File: src/Dictionaries/HashedDictionaryCollectionType.h [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+/// It turns out that HashMap can outperform google::sparse_hash_map in case of
+/// the structure size of not big, in terms of speed *and* memory. Even 99% of
+/// max load factor was faster then google::sparse_hash_map in my simple tests
+/// (1e9 UInt64 keys with UInt16 values, randomly distributed).
+///
+/// And not to mention very high allocator memory fragmentation in
+/// google::sparse_hash_map.
+///
+/// Based on HashTableGrowerWithPrecalculation
+class alignas(64) HashTableGrowerWithMaxLoadFactor
```
[the remark this answers is not in the document]
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionType.h:54]: > (cosmetic: could use a name that positions this class better within the existing system of growers, e.g. HashTableGrowerWithPrecalculationAndMaxLoadFactor

I thought that even for ClickHouse it will be too long, but if you have this mind I will update the code.

>Please also add a comment to HashTableGrower that it assumes a max load factor of 0.5.).

Ok

>I actually also think that HashTable.h would be a more fitting place for HashTableGrowerWithMaxLoadFactor (right after the existing growers).

Let's keep it here for now, since changes in `HashTable.h` requires tons of modules to recomple, once it will be stable it could be moved there.

File: src/Dictionaries/HashedDictionaryCollectionType.h [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    /// HashTableGrowerWithPrecalculation has 23
+    size_t max_size_degree_quadratic = 19;
+
+public:
+    static constexpr auto initial_count = 1ULL << initial_size_degree;
+
+    /// If collision resolution chains are contiguous, we can implement erase operation by moving the elements.
+    static constexpr auto performs_linear_probing_with_single_step = true;
+
+    HashTableGrowerWithMaxLoadFactor() = default;
```
[the remark this answers is not in the document]
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionType.h:70]: Yes, because there is another ctor.

File: src/Dictionaries/HashedDictionaryCollectionType.h [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+/// Based on HashTableGrowerWithPrecalculation
+class alignas(64) HashTableGrowerWithMaxLoadFactor
+{
+    static constexpr size_t initial_size_degree = 8;
+    UInt8 size_degree = initial_size_degree;
+    size_t precalculated_mask = (1ULL << initial_size_degree) - 1;
+    size_t precalculated_max_fill = 1ULL << (initial_size_degree - 1);
+    float max_load_factor = 0.5;
+    /// HashTableGrowerWithPrecalculation has 23
+    size_t max_size_degree_quadratic = 19;
```
[the remark this answers is not in the document]
Comment (azat) [reply src/Dictionaries/HashedDictionaryCollectionType.h:62]: >a) if 19 was found experimentally (if so, let's add a comment) <- probably yes, as "23" in HTGrowerWithPrecalculation looks equally "random"

This is just to decrease memory usage slightly and I doubt that there is huge difference (in terms of speed) between 23 and 19, especially for dictionaries (since they are not that frequently loaded).

>for symmetry, the corresponding constant in HashTableGrowerWithPrecalculation could be pulled out as max_size_degree.

OK, will do (though I think that touching extra bits in the patches makes review process harder.)

File: docs/en/sql-reference/dictionaries/index.md [hunk 12]
```diff
@@ -284,6 +284,12 @@ Configuration example:
          10000 is good balance between memory and speed.
          Even for 10e10 elements and can handle all the load without starvation. -->
     <shard_load_queue_backlog>10000</shard_load_queue_backlog>
+    <!-- Maximum load factor of the hash table, greater values decreases memory
```
[the remark this answers is not in the document]
Comment (azat) [reply docs/en/sql-reference/dictionaries/index.md]: This block was slightly out of sync in other places, so I've made it slightly better - 90639c283f425a7ad438aeb13f725c55d816d858

File: src/Common/HashTable/PackedHashMap.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (azat) [reply src/Common/HashTable/PackedHashMap.h]: [`06e8bd6` (#49380)](https://github.com/ClickHouse/ClickHouse/pull/49380/commits/06e8bd63b013d19b67548e99f3e5731226439e9f)

File: src/Dictionaries/HashedDictionaryCollectionTraits.h (hunk 3 above)
Comment (rschu1ze) [reply src/Dictionaries/HashedDictionaryCollectionTraits.h:25]: Okay, let's keep it as is.

**Review by rschu1ze — APPROVED** [review]
Lgtm, thanks. Will merge when the tests are green (will be OOO one day, so likely tomorrow afternoon CET).

**Comment by azat** [comment]
Performance tests looks [good](https://s3.amazonaws.com/clickhouse-test-reports/49380/5abc1a1c68ee204c9024493be1d19835cf5630f7/performance_comparison_[1_4]/report.html) even better then [before](https://s3.amazonaws.com/clickhouse-test-reports/49380/3daa1d880d4874993afc327f044f7587848aab6c/performance_comparison_%5B1_4%5D/report.html) (either it is clang-16 or it is just marginal)

Test failures not related:

>[Integration tests (release) [3/4] — fail: 5, passed: 499, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/49380/5abc1a1c68ee204c9024493be1d19835cf5630f7/integration_tests__release__[3_4].html)

- `test_distributed_load_balancing` is flaky will take a look

>[Unit tests (asan) — PASSED record not found.](https://s3.amazonaws.com/clickhouse-test-reports/49380/5abc1a1c68ee204c9024493be1d19835cf5630f7/unit_tests__asan_.html)

- #49908

>[Stress test (ubsan) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/49380/5abc1a1c68ee204c9024493be1d19835cf5630f7/stress_test__ubsan_.html)

```
 [ 2147 ] {6e68506e-d875-4199-98f6-425cce95e8af} <Fatal> : Logical error: \'Not enough space to add f57f4abfb9d2ad8c532b100a96b65bb2:0 with size 767: current size: 134216962/134217728\'.
```

- #49553

**Comment by azat** [comment]
Check failures are unrelated, performance test fails because float attributes added only in this PR for dictionaries and the tests:

>Stateless tests (tsan, s3 storage) [3/5] — fail: 1, passed: 1067, skipped: 34

- `02481_async_insert_race_long` - is flaky

**Comment by tavplubix** [comment]
It breaks performance tests, the quick fix did not help, so let's revert
